public/js/utils/sources-tree.js

```diff
@@ -161,6 +160,26 @@ function addToTree(tree: any, source: TmpSource) {
 }
 
 /**
+ * Look at the nodes in the source tree, and determine the index of where to
+ * insert a new node.
+ */
+function determineFileSortOrder(nodes:array, pathPart:string,
+                                isLastPart:boolean) {
+  const partIsDir = !isLastPart || pathPart.indexOf(".") === -1;
+
+  return nodes.findIndex(node => {
+    if (node.name === "(index)") {
+      return false;
+    }
+    const nodeIsDir = nodeHasChildren(node);
+    if (partIsDir) {
+      return nodeIsDir ? node.name.localeCompare(pathPart) >= 0 : true;
+    }
+    return nodeIsDir ? false : node.name.localeCompare(pathPart) >= 0;
```

Comment:
I think this can be re-worked to be clearer. I'm 50% confident this will work, so it's a good thing you have tests.

What I like about this though is that you're saying if both the part and node are dir or not dir, then do a comparison, otherwise and them. ... i don't know, definitely needs some comments here :)

```js
if (partIsDir && nodeIsDir) || !(partIsDir && nodeIsDir) {
  return node.name.localeCompare(pathPart);
}  else {
   return partIsDir && nodeIsDir;
}
```